tls/s2n_ktls.c

```diff
@@ -15,11 +15,176 @@
 
 #include "tls/s2n_ktls.h"
 
+#include <sys/socket.h>
+
+#include "utils/s2n_socket.h"
+
+/* These variables are used to disable ktls mechanisms during testing. */
+static bool disable_ktls_socket_config_for_testing = false;
+
 bool s2n_ktls_is_supported_on_platform()
 {
-#if defined(__linux__)
+#if S2N_KTLS_SUPPORTED
     return true;
 #else
     return false;
 #endif
 }
+
+S2N_RESULT s2n_ktls_validate(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(conn->secure);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite->record_alg);
+    const struct s2n_cipher *cipher = conn->secure->cipher_suite->record_alg->cipher;
+    RESULT_ENSURE_REF(cipher);
+
+    /* kTLS enable should only be called once the handshake has completed. */
+    if (!is_handshake_complete(conn)) {
+        RESULT_BAIL(S2N_ERR_KTLS_HANDSHAKE_NOT_COMPLETE);
+    }
+
+    /* TODO support TLS 1.3
+     *
+     * TLS 1.3 support requires sending the KeyUpdate message when the cryptographic
+     * KeyLimits are met. However, this is currently only possible by applying a
+     * kernel patch to support this functionality.
+     */
+    RESULT_ENSURE(conn->actual_protocol_version == S2N_TLS12, S2N_ERR_KTLS_UNSUPPORTED_CONN);
+
+    /* Check if the cipher supports kTLS */
+    RESULT_ENSURE(cipher->ktls_supported, S2N_ERR_KTLS_UNSUPPORTED_CONN);
+
+    /* kTLS I/O functionality is managed by s2n-tls. kTLS cannot be enabled if the
+     * application sets custom I/O (managed_send_io == false means application has
+     * set custom I/O).
+     */
+    switch (ktls_mode) {
+        case S2N_KTLS_MODE_SEND:
+            RESULT_ENSURE_REF(conn->send_io_context);
+            RESULT_ENSURE(conn->managed_send_io, S2N_ERR_KTLS_MANAGED_IO);
+
+            /* The output stuffer should be empty before enabling kTLS. */
+            RESULT_ENSURE(s2n_stuffer_data_available(&conn->out) == 0, S2N_ERR_RECORD_STUFFER_NEEDS_DRAINING);
+
+            break;
+        case S2N_KTLS_MODE_RECV:
+            RESULT_ENSURE_REF(conn->recv_io_context);
+            RESULT_ENSURE(conn->managed_recv_io, S2N_ERR_KTLS_MANAGED_IO);
+
+            /* The input stuffer should be empty before enabling kTLS. */
+            RESULT_ENSURE(s2n_stuffer_data_available(&conn->in) == 0, S2N_ERR_RECORD_STUFFER_NEEDS_DRAINING);
+
+            break;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_retrieve_file_descriptor(struct s2n_connection *conn, s2n_ktls_mode ktls_mode, int *fd)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(fd);
+
+    if (ktls_mode == S2N_KTLS_MODE_RECV) {
+        /* retrieve the receive fd */
+        RESULT_ENSURE_REF(conn->recv_io_context);
+        const struct s2n_socket_read_io_context *peer_socket_ctx = conn->recv_io_context;
```

Comment:
Why is this variable called peer_socket_ctx? What does that mean?

Reply:
Actually this can be replaced with `s2n_connection_get_read_fd(conn, fd)`. Its essentially getting the read file descriptor from the read context.

Since have to be managed_io the underlying context should be a `s2n_socket_read_io_context`.